File: agent-packs/eval-pilot/engine/src/evalpilot/discovery.py
"""Discover agents and skills across Copilot's supported layouts.

A target repository may declare agents and skills in several places. The
original monorepo harness only knew about ``agent-packs/<pack>/.github/...``;
this generalised discovery understands every layout the Copilot CLI itself
loads from, so evalpilot works in an arbitrary repo:

Agents (``<name>.agent.md``):
  * ``.github/agents/`` (repo-level custom agents)
  * ``<plugin>/agents/`` (plugin-root layout, e.g. ``agent-packs/<pack>/agents``)
  * ``.github/agents/`` *inside* a pack (monorepo legacy)

Skills (``<name>/SKILL.md``):
  * ``.github/skills/<name>/`` (repo-level / shared)
  * ``<plugin>/skills/<name>/`` (plugin-root layout)

The discovery functions return small descriptors so the workspace stager
and the CLI can present "what can I evaluate here?" without re-implementing
the walk.
"""

from __future__ import annotations

import dataclasses
from pathlib import Path
from typing import Iterable, Optional

from .config import find_repo_root


# Directories we never descend into while discovering (build output, vcs,
# virtualenvs, the engine's own scratch).
_PRUNE = {
    ".git", ".hg", ".svn", "node_modules", "__pycache__", ".venv", "venv",
    ".pytest_cache", ".mypy_cache", "dist", "build", ".tox", "_runs",
    "_logs", "_metrics",
}


@dataclasses.dataclass(frozen=True)
class AgentInfo:
    """A discovered custom agent."""

    name: str
    """Agent name (filename stem, e.g. ``eval-judge`` for ``eval-judge.agent.md``)."""

    path: Path
    """Absolute path to the ``.agent.md`` file."""

    agents_dir: Path
    """The ``agents/`` (or ``.github/agents/``) directory containing it."""

    plugin_root: Optional[Path]
    """Plugin root if this agent belongs to a plugin (manifest sibling), else None."""


@dataclasses.dataclass(frozen=True)
class SkillInfo:
    """A discovered skill."""

    name: str
    """Skill name (the ``SKILL.md`` parent directory name)."""

    path: Path
    """Absolute path to the skill directory (the one containing ``SKILL.md``)."""

    skills_dir: Path
    """The ``skills/`` (or ``.github/skills/``) directory containing it."""

    plugin_root: Optional[Path]
    """Plugin root if this skill belongs to a plugin, else None."""


def _iter_dirs(root: Path, name: str) -> Iterable[Path]:
    """Yield every directory named ``name`` under ``root`` (pruned)."""
    if not root.exists():
        return
    for dirpath, dirnames, _ in _walk(root):
        # prune in place
        dirnames[:] = [d for d in dirnames if d not in _PRUNE]
        if dirpath.name == name:
            yield dirpath


def _walk(root: Path):
    """A small ``os.walk``-like generator yielding ``Path`` dirpaths."""
    import os

    for dirpath, dirnames, filenames in os.walk(root):
        yield Path(dirpath), dirnames, filenames


def _plugin_root_for(start: Path) -> Optional[Path]:
    """Return the nearest ancestor of ``start`` containing a ``plugin.json``."""
    for parent in [start, *start.parents]:
        if (parent / "plugin.json").is_file():
            return parent
    return None


def discover_agents(repo_root: Optional[Path] = None) -> list[AgentInfo]:
    """Return all custom agents declared anywhere in the repo."""
    repo_root = repo_root or find_repo_root()
    seen: dict[Path, AgentInfo] = {}
    for agents_dir in _iter_dirs(repo_root, "agents"):
        for f in sorted(agents_dir.glob("*.agent.md")):
            info = AgentInfo(
                name=f.name[: -len(".agent.md")],
                path=f.resolve(),
                agents_dir=agents_dir.resolve(),
                plugin_root=_plugin_root_for(agents_dir),
            )
            seen[info.path] = info
    return sorted(seen.values(), key=lambda a: (a.name, str(a.path)))


def discover_skills(repo_root: Optional[Path] = None) -> list[SkillInfo]:
    """Return all skills declared anywhere in the repo."""
    repo_root = repo_root or find_repo_root()
    seen: dict[Path, SkillInfo] = {}
    for skills_dir in _iter_dirs(repo_root, "skills"):
        for skill_md in sorted(skills_dir.glob("*/SKILL.md")):
            skill_dir = skill_md.parent
            info = SkillInfo(
                name=skill_dir.name,
                path=skill_dir.resolve(),
                skills_dir=skills_dir.resolve(),
                plugin_root=_plugin_root_for(skills_dir),
            )
            seen[info.path] = info
    return sorted(seen.values(), key=lambda s: (s.name, str(s.path)))


def find_agent(name: str, repo_root: Optional[Path] = None) -> AgentInfo:
    """Return the single agent named ``name`` or raise."""
    matches = [a for a in discover_agents(repo_root) if a.name == name]
    if len(matches) == 1:
        return matches[0]
    if not matches:
        available = ", ".join(a.name for a in discover_agents(repo_root)) or "(none)"
        raise LookupError(f"No agent named {name!r}. Available: {available}")
    raise LookupError(
        f"Agent {name!r} is ambiguous across {len(matches)} locations: "
        f"{[str(m.path) for m in matches]}"
    )


def find_skill(name: str, repo_root: Optional[Path] = None) -> SkillInfo:
    """Return the single skill named ``name`` or raise."""
    matches = [s for s in discover_skills(repo_root) if s.name == name]
    if len(matches) == 1:
        return matches[0]
    if not matches:
        available = ", ".join(s.name for s in discover_skills(repo_root)) or "(none)"
        raise LookupError(f"No skill named {name!r}. Available: {available}")
    raise LookupError(
        f"Skill {name!r} is ambiguous across {len(matches)} locations: "
        f"{[str(m.path) for m in matches]}"
    )


__all__ = [
    "AgentInfo",
    "SkillInfo",
    "discover_agents",
    "discover_skills",
    "find_agent",
    "find_skill",
]

File: agent-packs/eval-pilot/engine-ts/src/discovery.ts
/**
 * Discover agents and skills across Copilot's supported layouts.
 *
 * Understands every layout the Copilot CLI itself loads from, so evalpilot
 * works in an arbitrary repo:
 *
 * Agents (`<name>.agent.md`):
 *   - `.github/agents/`      (repo-level custom agents)
 *   - `<plugin>/agents/`     (plugin-root layout)
 *
 * Skills (`<name>/SKILL.md`):
 *   - `.github/skills/<name>/`
 *   - `<plugin>/skills/<name>/`
 */

import { existsSync, readdirSync, statSync } from "node:fs";
import * as path from "node:path";
import { findRepoRoot } from "./config.js";

const PRUNE = new Set<string>([
  ".git",
  ".hg",
  ".svn",
  "node_modules",
  "__pycache__",
  ".venv",
  "venv",
  ".pytest_cache",
  ".mypy_cache",
  "dist",
  "build",
  ".tox",
  "_runs",
  "_logs",
  "_metrics",
]);

/** A discovered custom agent. */
export interface AgentInfo {
  /** Agent name (filename stem, e.g. `eval-judge`). */
  name: string;
  /** Absolute path to the `.agent.md` file. */
  path: string;
  /** The `agents/` directory containing it. */
  agents_dir: string;
  /** Plugin root if this agent belongs to a plugin, else null. */
  plugin_root: string | null;
}

/** A discovered skill. */
export interface SkillInfo {
  /** Skill name (the `SKILL.md` parent directory name). */
  name: string;
  /** Absolute path to the skill directory (containing `SKILL.md`). */
  path: string;
  /** The `skills/` directory containing it. */
  skills_dir: string;
  /** Plugin root if this skill belongs to a plugin, else null. */
  plugin_root: string | null;
}

/** Yield every directory named `name` under `root` (pruned). */
function* iterDirs(root: string, name: string): Generator<string> {
  if (!existsSync(root)) return;
  const stack: string[] = [root];
  while (stack.length) {
    const dir = stack.pop()!;
    let entries: import("node:fs").Dirent[];
    try {
      entries = readdirSync(dir, { withFileTypes: true });
    } catch {
      continue;
    }
    if (path.basename(dir) === name) yield dir;
    for (const e of entries) {
      if (!e.isDirectory()) continue;
      if (PRUNE.has(e.name)) continue;
      stack.push(path.join(dir, e.name));
    }
  }
}

/** Return the nearest ancestor of `start` containing a `plugin.json`. */
function pluginRootFor(start: string): string | null {
  let current = path.resolve(start);
  // eslint-disable-next-line no-constant-condition
  while (true) {
    const manifest = path.join(current, "plugin.json");
    if (existsSync(manifest) && statSync(manifest).isFile()) return current;
    const parent = path.dirname(current);
    if (parent === current) break;
    current = parent;
  }
  return null;
}

function listFiles(dir: string, suffix: string): string[] {
  try {
    return readdirSync(dir, { withFileTypes: true })
      .filter((e) => e.isFile() && e.name.endsWith(suffix))
      .map((e) => path.join(dir, e.name))
      .sort();
  } catch {
    return [];
  }
}

function listSubdirsWithSkillMd(dir: string): string[] {
  try {
    return readdirSync(dir, { withFileTypes: true })
      .filter((e) => e.isDirectory())
      .map((e) => path.join(dir, e.name))
      .filter((d) => existsSync(path.join(d, "SKILL.md")))
      .sort();
  } catch {
    return [];
  }
}

/** Return all custom agents declared anywhere in the repo. */
export function discoverAgents(repoRoot?: string): AgentInfo[] {
  const rr = repoRoot ?? findRepoRoot();
  const seen = new Map<string, AgentInfo>();
  for (const agentsDir of iterDirs(rr, "agents")) {
    for (const f of listFiles(agentsDir, ".agent.md")) {
      const info: AgentInfo = {
        name: path.basename(f).slice(0, -".agent.md".length),
        path: path.resolve(f),
        agents_dir: path.resolve(agentsDir),
        plugin_root: pluginRootFor(agentsDir),
      };
      seen.set(info.path, info);
    }
  }
  return [...seen.values()].sort(
    (a, b) => a.name.localeCompare(b.name) || a.path.localeCompare(b.path),
  );
}

/** Return all skills declared anywhere in the repo. */
export function discoverSkills(repoRoot?: string): SkillInfo[] {
  const rr = repoRoot ?? findRepoRoot();
  const seen = new Map<string, SkillInfo>();
  for (const skillsDir of iterDirs(rr, "skills")) {
    for (const skillDir of listSubdirsWithSkillMd(skillsDir)) {
      const info: SkillInfo = {
        name: path.basename(skillDir),
        path: path.resolve(skillDir),
        skills_dir: path.resolve(skillsDir),
        plugin_root: pluginRootFor(skillsDir),
      };
      seen.set(info.path, info);
    }
  }
  return [...seen.values()].sort(
    (a, b) => a.name.localeCompare(b.name) || a.path.localeCompare(b.path),
  );
}

/** Return the single agent named `name` or throw. */
export function findAgent(name: string, repoRoot?: string): AgentInfo {
  const all = discoverAgents(repoRoot);
  const matches = all.filter((a) => a.name === name);
  if (matches.length === 1) return matches[0]!;
  if (matches.length === 0) {
    const available = all.map((a) => a.name).join(", ") || "(none)";
    throw new Error(`No agent named '${name}'. Available: ${available}`);
  }
  throw new Error(
    `Agent '${name}' is ambiguous across ${matches.length} locations: ` +
      JSON.stringify(matches.map((m) => m.path)),
  );
}

/** Return the single skill named `name` or throw. */
export function findSkill(name: string, repoRoot?: string): SkillInfo {
  const all = discoverSkills(repoRoot);
  const matches = all.filter((s) => s.name === name);
  if (matches.length === 1) return matches[0]!;
  if (matches.length === 0) {
    const available = all.map((s) => s.name).join(", ") || "(none)";
    throw new Error(`No skill named '${name}'. Available: ${available}`);
  }
  throw new Error(
    `Skill '${name}' is ambiguous across ${matches.length} locations: ` +
      JSON.stringify(matches.map((m) => m.path)),
  );
}
